File: leetCode/longest_palindromic_substring.js

// LeetCode
// # 5
// Medium


/*

Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

Example 1:

Input: "babad"
Output: "bab"
Note: "aba" is also a valid answer.

Example 2:

Input: "cbbd"
Output: "bb"

*/


const longestPalindrome = function(s) {
    if (!s || s.length === 1) return s;
    let longest = '';

    if (s[0] === s[s.length - 1]) {
		if (isPalindrome(s)) return s;
    }

    for (let i = 0; i < s.length; i += 1) {
    	for (let n = i; n < s.length; n += 1) {
    		if (s[i] === s[n]) {
    			let sub = s.slice(i, n + 1);
    			if (isPalindrome(sub) && sub.length > longest.length) {
    				longest = sub;
    			};
    		};
    	};
    };

    return longest;
};


const isPalindrome = (s) => {
  let i = 0;
  let n = s.length - 1;
  
  while (i < n) {
    i += 1;
    n -= 1;

    if (s[i] !== s[n]) return false;
  };
  
  return true;
};

File: leetCode/py3/longest_palindromic_substring.py

# LeetCode
# 5
# Medium

# Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

# Example 1:

# Input: "babad"
# Output: "bab"
# Note: "aba" is also a valid answer.

# Example 2:

# Input: "cbbd"
# Output: "bb"



class Solution:
    def longestPalindrome(self, s: str) -> str:
        '''
        loop thru the string
        on each char, check both left and right
        if equal, fan out
        '''
        if s == '' or len(s) == 1: return s
        if s == s[::-1]: return s

        longest = ''
        for i in range(len(s)):
            for n in range(2):
                left = i
                right = i + n

                while left >= 0 and right < len(s) and s[left] == s[right]:
                    left -= 1
                    right += 1

                if right - left - 1 > len(longest):
                    longest = s[left + 1 : right]

        return longest

